src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs

```diff
@@ -25,38 +25,48 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceScopeFac
                 throw new ArgumentNullException(nameof(scopeFactory));
             }
 
-            _scopeFactory = scopeFactory;
             _next = next;
+            _scopeFactory = scopeFactory;
         }
 
         public async Task Invoke(HttpContext httpContext)
         {
             if (httpContext == null)
             {
-                throw new ArgumentNullException(nameof(httpContext));
+                // Defer Exception creation and throw to non-returning throw helper
+                ThrowArgumentNullExceptionHttpContext();
             }
 
-            var existingFeature = httpContext.Features.Get<IServiceProvidersFeature>();
+            // local cache for virtual disptach result
+            var features = httpContext.Features;
+            var existingFeature = features.Get<IServiceProvidersFeature>();
 
-            // All done if request services is set
+            // All done if RequestServices is set
             if (existingFeature?.RequestServices != null)
             {
                 await _next.Invoke(httpContext);
                 return;
             }
```

Comment:
My immediate thought is to care this up into a non-async path and an async path, but my guess is that we'd always take the async path.